### PR TITLE
Simplify item rendering loop

### PR DESCRIPTION
**Pull request by tronical**
Instead of using the generic visitor, call `visit_children_item` directly on the Component.

This way we avoid the RefCell for the renderer and this will make it
easier to introduce items that decide to render the children themselves, with
their own property tracker.

### CODE REVIEW

**Comment by tronical** [comment]
This needs a careful review. If this is the right approach, then I can also apply it to the other user of `visit_items_with_post_visit`, the mouse event delivery code.

File: internal/core/item_rendering.rs [hunk 1]
```diff
@@ -95,6 +96,43 @@ pub(crate) fn is_clipping_item(item: Pin<ItemRef>) -> bool {
         || ItemRef::downcast_pin::<Clip>(item).is_some()
 }
 
+fn render_item_children(
+    renderer: &mut dyn ItemRenderer,
+    component: &ComponentRc,
+    index: isize,
```
Comment (ogoffart) [inline internal/core/item_rendering.rs:102]: That's an ItemRc.

File: internal/core/item_rendering.rs [hunk 2]
```diff
@@ -95,6 +96,43 @@ pub(crate) fn is_clipping_item(item: Pin<ItemRef>) -> bool {
...
+
+            let (do_draw, item_geometry) = renderer.filter_item(item);
+
+            let item_origin = item_geometry.origin;
+            renderer.translate(item_origin.x, item_origin.y);
+
+            // Don't render items that are clipped, with the exception of the Clip or Flickable since
+            // they themselves clip their content.
+            if do_draw
+               || is_clipping_item(item) // HACK, the geometry of the box shadow does not include the shadow, because when the shadow is the root for repeated elements it would translate the children
```
Comment (ogoffart) [inline internal/core/item_rendering.rs:116]: The comment apply to the next love

File: internal/core/item_rendering.rs [hunk 3]
```diff
@@ -95,6 +96,43 @@ pub(crate) fn is_clipping_item(item: Pin<ItemRef>) -> bool {
...
+               || ItemRef::downcast_pin::<BoxShadow>(item).is_some()
+            {
+                item.as_ref().render(&mut (renderer as &mut dyn ItemRenderer));
+            }
+
+            let result = render_item_children(renderer, component, index as isize);
+
+            renderer.restore_state();
+
```
Comment (ogoffart) [inline internal/core/item_rendering.rs:125]: Why so much empty lines 😃

**Review by ogoffart — COMMENTED** [review]
Looks good

File: internal/core/item_rendering.rs (hunk 1 above)
Comment (tronical) [reply internal/core/item_rendering.rs:102]: That's what I thought at first, too, but it turns out that's not the case for the very first call, where the index is `-1`. I could work around it by casting, but that feels more like a hack.

File: internal/core/item_rendering.rs (hunk 2 above)
Comment (tronical) [reply internal/core/item_rendering.rs:116]: Oops yes. Oddly `cargo fmt` gives up on this closure. Not sure why.

**Review by hunger — COMMENTED** [review]
Format looks strange (at least in the github app)... I understand that you copy/pasted it in this way, but a reformat would be nice.

Apart from that the patch looks ok to me;-)

File: internal/core/item_rendering.rs (hunk 3 above)
Comment (tronical) [reply internal/core/item_rendering.rs:125]: Pure generosity ;-)

(I like the visual separation between unrelated calls, but it's maybe extreme here - fixed now)

**Comment by tronical** [comment]
> Format looks strange (at least in the github app)... I understand that you copy/pasted it in this way, but a reformat would be nice.

I've tried reformatting multiple times, but it's not changing the code inside the closure.

**Review by ogoffart — APPROVED** [review]

**Pull request merged**